src/VisualStudio/Core/Def/ProjectSystem/BrokeredService/WorkspaceProjectFactory.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Remote.ProjectSystem;
+using Microsoft.ServiceHub.Framework;
+
+namespace Microsoft.VisualStudio.LanguageServices.ProjectSystem.BrokeredService
+{
+    internal class WorkspaceProjectFactory : IWorkspaceProjectFactory
+    {
+        public const string ServiceName = "WorkspaceProjectFactory";
+        public static readonly ServiceDescriptor ServiceDescriptor = ServiceDescriptor.CreateInProcServiceDescriptor(ServiceDescriptors.ComponentName, ServiceName, suffix: "", ServiceDescriptors.GetFeatureDisplayName);
+
+        private readonly IWorkspaceProjectContextFactory _workspaceProjectContextFactory;
+
+        // For the sake of the in-proc implementation here, we're going to build this atop IWorkspaceProjectContext so semantics are preserved
+        // for a few edge cases. Once the project system has moved onto this directly, we can flatten the implementations out.
+        public WorkspaceProjectFactory(IWorkspaceProjectContextFactory workspaceProjectContextFactory)
+        {
+            _workspaceProjectContextFactory = workspaceProjectContextFactory;
+        }
+
+        public async Task<IWorkspaceProject> CreateAndAddProjectAsync(WorkspaceProjectCreationInfo creationInfo, CancellationToken cancellationToken)
+        {
+            var project = await _workspaceProjectContextFactory.CreateProjectContextAsync(
+                Guid.NewGuid(), // TODO: figure out some other side-channel way of communicating this
```

Comment:
We only use the GUID for specific interactions with the table control, we might be able to get rid of this. In any case, it has no reason to exist on the cross-product interface.